src/librbd/AioImageRequest.cc

```diff
@@ -724,10 +733,155 @@ void AioImageFlush<I>::send_image_cache_request() {
+                                                     bufferlist *bl) {
+  bool may_writesame = true;
+
+  for (auto q = object_extent.buffer_extents.begin();
+       q != object_extent.buffer_extents.end(); ++q) {
+    if (!(q->first % m_data_len == 0 && q->second % m_data_len == 0)) {
+      may_writesame = false;
+      break;
+    }
+  }
+
+  if (may_writesame) {
+    bl->append(m_buf, m_data_len);
+    return true;
+  }
+
+  uint64_t sub_off, sub_len, extent_left;
+  for (auto q = object_extent.buffer_extents.begin();
+       q != object_extent.buffer_extents.end(); ++q) {
+    sub_off = q->first % m_data_len;
+    sub_len = m_data_len - sub_off;
+    extent_left = q->second;
+    while (extent_left >= sub_len) {
+      bl->append(m_buf + sub_off, sub_len);
+      extent_left -= sub_len;
+      if (sub_off) {
+	sub_off = 0;
+	sub_len = m_data_len;
+      }
+    }
+    if (extent_left)
+      bl->append(m_buf + sub_off, extent_left);
+  }
+  return false;
+}
+
+template <typename I>
+uint64_t AioImageWriteSame<I>::append_journal_event(
+    const AioObjectRequests &requests, bool synchronous) {
+  I &image_ctx = this->m_image_ctx;
+
+  bufferlist bl;
+  bl.append(m_buf, m_data_len);
+  uint64_t tid;
+  assert(!this->m_image_extents.empty());
+
+  for (auto &extent : this->m_image_extents) {
+    journal::EventEntry event_entry(journal::AioWriteSameEvent(extent.first,
+                                                               extent.second,
+                                                               bl));
+    tid = image_ctx.journal->append_io_event(std::move(event_entry), requests,
+                                             extent.first, extent.second, synchronous);
+  }
+
+  if (image_ctx.object_cacher == NULL) {
+    AioCompletion *aio_comp = this->m_aio_comp;
+    aio_comp->associate_journal_event(tid);
+  }
+  return tid;
+}
+
+template <typename I>
+void AioImageWriteSame<I>::send_image_cache_request() {
+  I &image_ctx = this->m_image_ctx;
+  assert(image_ctx.image_cache != nullptr);
+
+  AioCompletion *aio_comp = this->m_aio_comp;
+  aio_comp->set_request_count(this->m_image_extents.size());
+  C_AioRequest *req_comp = new C_AioRequest(aio_comp);
+  for (auto &extent : this->m_image_extents) {
+    image_ctx.image_cache->aio_writesame(extent.first, extent.first,
+                                         m_buf, m_data_len,
+                                         m_op_flags, req_comp);
+  }
+}
+
+template <typename I>
+void AioImageWriteSame<I>::send_object_cache_requests(const ObjectExtents &object_extents,
+                                                      uint64_t journal_tid) {
+  I &image_ctx = this->m_image_ctx;
+  for (auto p = object_extents.begin(); p != object_extents.end(); ++p) {
+    const ObjectExtent &object_extent = *p;
+
+    bufferlist bl;
+    if (assemble_writesame_extent(object_extent, &bl)) {
+      for (int i = 0; i < object_extent.length / m_data_len - 1; i++)
```

Comment:
Ack -- missed that the ```bl``` is passed into the ```assemble_writesame_extent``` method. A short comment would be helpful.